### PR TITLE
SDN-4783: Add e2e tests for NetworkDiagnosticsConfig featuregate

### PR DESCRIPTION
**Pull request by kyrtapz**
This is blocked by https://github.com/openshift/cluster-network-operator/pull/2339

### CODE REVIEW

File: test/extended/networking/network_diagnostics.go [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+
+	g.AfterEach(func(ctx context.Context) {
+		// Reset network diagnostics config
+		netConfigApply := applyconfigv1.Network(clusterConfig).WithSpec(
+			applyconfigv1.NetworkSpec().WithNetworkDiagnostics(nil),
+		)
+		_, err := oc.AdminConfigClient().ConfigV1().Networks().Apply(ctx, netConfigApply,
+			metav1.ApplyOptions{FieldManager: fieldManager, Force: true})
+		o.Expect(err).NotTo(o.HaveOccurred())
+	})
```
Comment (ricky-rav) [inline test/extended/networking/network_diagnostics.go:39]: Would it make sense to also check this state, either in a separate test or before each test begins?

File: test/extended/networking/network_diagnostics.go [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		o.Eventually(func() bool {
+			// Should not affect the Progressing condition of network.operator
+			oper, err := oc.AdminOperatorClient().OperatorV1().Networks().Get(ctx, clusterConfig, metav1.GetOptions{})
+			if err != nil {
+				framework.Logf("Error getting cluster operator: %v", err)
+				return false
+			}
+			for _, operCondition := range oper.Status.Conditions {
```
Comment (ricky-rav) [inline test/extended/networking/network_diagnostics.go:232]: Should we also check that CNO isn't progressing in the other tests where `meta.IsStatusConditionFalse`?

File: test/extended/networking/network_diagnostics.go [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+					),
+			),
+		)
+		_, err := oc.AdminConfigClient().ConfigV1().Networks().Apply(ctx, netConfigApply,
+			metav1.ApplyOptions{FieldManager: fieldManager, Force: true})
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		o.Eventually(func() bool {
+			// Should not affect the Progressing condition of network.operator
+			oper, err := oc.AdminOperatorClient().OperatorV1().Networks().Get(ctx, clusterConfig, metav1.GetOptions{})
```
Comment (ricky-rav) [inline test/extended/networking/network_diagnostics.go:227]: would it be redundant to also check for no source pods here?

**Review by ricky-rav — COMMENTED** [review]
Overall looks good, I left a few questions.

File: test/extended/networking/network_diagnostics.go (hunk 1 above)
Comment (kyrtapz) [reply test/extended/networking/network_diagnostics.go:40]: I've added the test to verify the default behaviour.

File: test/extended/networking/network_diagnostics.go (hunk 2 above)
Comment (kyrtapz) [reply test/extended/networking/network_diagnostics.go:273]: I don't think it is required in other cases as nothing impacts pod deployment.

File: test/extended/networking/network_diagnostics.go (hunk 3 above)
Comment (kyrtapz) [reply test/extended/networking/network_diagnostics.go:268]: The pods will be there but it will be stack in scheduling as it's waiting for a valid node.

**Comment by ricky-rav** [comment]
/lgtm
Thanks, @kyrtapz!

**Review by trozet — APPROVED** [review]

**Comment by kyrtapz** [comment]
CI seems to be back
/retest

**Comment by kyrtapz** [comment]
@dgoodwin are you ok with the cliet-go bump? It looks like the bot needs your blessing

**Comment by kyrtapz** [comment]
Made the tests ordered so they do not run in parallel with each other:
```
Ordered is a decorator that allows you to mark a container as ordered.  Specs in the container will always run in the order they appear.
They will never be randomized and they will never run in parallel with one another, though they may run in parallel with other specs.
```

**Comment by kyrtapz** [comment]
> I believe that moves them to the serial suite and thus they will then only run in serial jobs. I would not expect them to be running in e2e-gcp-ovn-techpreview anymore. How long does this set of tests take to run? Just want to ask in terms of time added to the suite.

Each test has a timeout of 3mins, from my testing each test doesn't take more than a minute.

**Comment by dgoodwin** [comment]
> > I believe that moves them to the serial suite and thus they will then only run in serial jobs. I would not expect them to be running in e2e-gcp-ovn-techpreview anymore. How long does this set of tests take to run? Just want to ask in terms of time added to the suite.
> 
> Each test has a timeout of 3mins, from my testing each test doesn't take more than a minute.

That should be fine in the serial suite.

**Comment by kyrtapz** [comment]
Rebased and ran `make update`
/test e2e-gcp-ovn-techpreview-serial

**Comment by kyrtapz** [comment]
Network diagnostics tests passed previously, lets try again
/test e2e-gcp-ovn-techpreview-serial

**Comment by kyrtapz** [comment]
Network diagnostics tests passed, the unrelated failure is already [tracked](https://issues.redhat.com/browse/OCPBUGS-30224)

**Comment by kyrtapz** [comment]
Had to re-vendor after rebasing to master.

**Pull request merged**